File: apps/mobile/client-apple/VettaKit/Sources/VettaKit/App/ChatTurns.swift
import Foundation

/// One step of an agent's work, shown as a row inside a folded work group.
public enum WorkStep: Equatable, Identifiable, Sendable {
	case thinking(id: String, text: String)
	case tool(ToolCard)

	public var id: String {
		switch self {
		case let .thinking(id, _): id
		case let .tool(card): card.id
		}
	}

	public var pending: Bool {
		if case let .tool(card) = self { return card.status == .running || card.status == .generating }
		return false
	}
}

/// A piece of a turn in display order: thinking and tools fold together until
/// the agent writes text, which closes the group (as the desktop's work stages do).
public enum TurnSegment: Equatable, Identifiable, Sendable {
	case work(id: String, steps: [WorkStep])
	case text(id: String, text: String)
	/// `count` > 1 when the same failure repeated, e.g. over automatic retries.
	case error(id: String, message: String, count: Int)

	public var id: String {
		switch self {
		case let .work(id, _), let .text(id, _), let .error(id, _, _): id
		}
	}
}

/// Everything the agent did between two user messages, merged into one turn.
public struct AgentTurn: Equatable, Identifiable, Sendable {
	public var id: String
	public var segments: [TurnSegment]
	public var streaming: Bool
	public var startedAt: Double?

	/// The closing answer, for the copy button: the text after the last work group.
	public var conclusion: String {
		var texts: [String] = []
		for segment in segments.reversed() {
			switch segment {
			case let .text(_, text): texts.insert(text, at: 0)
			case .error: continue
			case .work: return texts.joined(separator: "\n\n")
			}
		}
		return texts.joined(separator: "\n\n")
	}

	/// What the agent is doing right now, for the live group title.
	public var activity: WorkStep? {
		guard streaming, case let .work(_, steps) = segments.last else { return nil }
		return steps.last(where: \.pending) ?? steps.last
	}
}

public enum ChatBlock: Equatable, Identifiable, Sendable {
	case user(id: String, text: String, at: Double?, attachments: [TranscriptAttachment])
	case marker(id: String, text: String, at: Double?)
	case turn(AgentTurn)

	public var id: String {
		switch self {
		case let .user(id, _, _, _), let .marker(id, _, _): id
		case let .turn(turn): turn.id
		}
	}
}

public enum ChatTurns {
	/// Merges consecutive assistant items into one turn. A user message or a
	/// marker (e.g. compaction) ends the turn; errors stay inside it. `waiting`
	/// means the session is working: a turn with nothing to show yet still
	/// appears, so a sent message never sits there without feedback.
	public static func build(_ items: [TranscriptItem], waiting: Bool = false) -> [ChatBlock] {
		var blocks: [ChatBlock] = []
		for item in items {
			switch item {
			case let .user(id, text, at, attachments):
				blocks.append(.user(id: id, text: text, at: at, attachments: attachments))
			case let .marker(id, text, at):
				blocks.append(.marker(id: id, text: text, at: at))
			case let .assistant(reply):
				var turn: AgentTurn
				if case let .turn(existing) = blocks.last {
					turn = existing
					blocks.removeLast()
				} else {
					turn = AgentTurn(id: reply.id, segments: [], streaming: false, startedAt: reply.at)
				}
				append(reply, to: &turn)
				turn.streaming = reply.streaming
				blocks.append(.turn(turn))
			}
		}
		if waiting {
			if case var .turn(turn) = blocks.last {
				turn.streaming = true
				blocks[blocks.count - 1] = .turn(turn)
			} else {
				blocks.append(.turn(AgentTurn(id: "pending-turn", segments: [], streaming: true, startedAt: nil)))
			}
		}
		return blocks
	}

	private static func append(_ reply: AssistantTurn, to turn: inout AgentTurn) {
		var steps: [WorkStep] = []
		if !reply.thinking.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
			steps.append(.thinking(id: "\(reply.id)-thinking", text: reply.thinking))
		}
		steps += reply.tools.map(WorkStep.tool)
		let text = reply.text.trimmingCharacters(in: .whitespacesAndNewlines)
		if !steps.isEmpty || !text.isEmpty {
			// New output means the attempts that failed before it were retried
			// successfully; like the desktop, those failures are dropped.
			while case .error = turn.segments.last { turn.segments.removeLast() }
		}
		if !steps.isEmpty {
			if case let .work(id, previous) = turn.segments.last {
				turn.segments[turn.segments.count - 1] = .work(id: id, steps: previous + steps)
			} else {
				turn.segments.append(.work(id: "\(reply.id)-work", steps: steps))
			}
		}
		if !text.isEmpty {
			turn.segments.append(.text(id: "\(reply.id)-text", text: reply.text))
		}
		if let error = reply.error {
			if case let .error(id, message, count) = turn.segments.last, message == error {
				turn.segments[turn.segments.count - 1] = .error(id: id, message: message, count: count + 1)
			} else {
				turn.segments.append(.error(id: "\(reply.id)-error", message: error, count: 1))
			}
		}
	}
}

File: apps/mobile/client-android/shared/src/commonMain/kotlin/org/vetta/android/domain/work/ChatTurns.kt
package org.vetta.android.domain.work

import org.vetta.android.domain.remote.AssistantTurn
import org.vetta.android.domain.remote.ToolCard
import org.vetta.android.domain.remote.ToolCardStatus
import org.vetta.android.domain.remote.TranscriptAttachment
import org.vetta.android.domain.remote.TranscriptItem

/** One step of an agent's work, shown as a row inside a folded work group. */
sealed interface WorkStep {
    val id: String

    data class Thinking(override val id: String, val text: String) : WorkStep

    data class Tool(val card: ToolCard) : WorkStep {
        override val id: String
            get() = card.toolCallId
    }

    val pending: Boolean
        get() = this is Tool && (card.status == ToolCardStatus.Running || card.status == ToolCardStatus.Generating)
}

/**
 * A piece of a turn in display order: thinking and tools fold together until
 * the agent writes text, which closes the group (as the desktop's work stages do).
 */
sealed interface TurnSegment {
    val id: String

    data class Work(override val id: String, val steps: List<WorkStep>) : TurnSegment

    data class Text(override val id: String, val text: String) : TurnSegment

    /** `count` > 1 when the same failure repeated, e.g. over automatic retries. */
    data class Error(override val id: String, val message: String, val count: Int) : TurnSegment
}

/** Everything the agent did between two user messages, merged into one turn. */
data class AgentTurn(
    val id: String,
    val segments: List<TurnSegment>,
    val streaming: Boolean,
    val startedAt: Long?,
) {
    /** The closing answer, for the copy button: the text after the last work group. */
    val conclusion: String
        get() {
            val texts = ArrayDeque<String>()
            for (segment in segments.asReversed()) {
                when (segment) {
                    is TurnSegment.Text -> texts.addFirst(segment.text)
                    is TurnSegment.Error -> continue
                    is TurnSegment.Work -> break
                }
            }
            return texts.joinToString("\n\n")
        }

    /** What the agent is doing right now, for the live group title. */
    val activity: WorkStep?
        get() {
            if (!streaming) return null
            val work = segments.lastOrNull() as? TurnSegment.Work ?: return null
            return work.steps.lastOrNull { it.pending } ?: work.steps.lastOrNull()
        }
}

sealed interface ChatBlock {
    val id: String

    data class User(
        override val id: String,
        val text: String,
        val at: Long?,
        val attachments: List<TranscriptAttachment>,
    ) : ChatBlock

    data class Marker(override val id: String, val text: String, val at: Long?) : ChatBlock

    data class Turn(val turn: AgentTurn) : ChatBlock {
        override val id: String
            get() = turn.id
    }
}

object ChatTurns {
    /** The turn shown while a session works but has written nothing yet. */
    const val PENDING_TURN_ID = "pending-turn"

    /**
     * Merges consecutive assistant items into one turn. A user message or a
     * marker (e.g. compaction) ends the turn; errors stay inside it. `waiting`
     * means the session is working: a turn with nothing to show yet still
     * appears, so a sent message never sits there without feedback.
     */
    fun build(items: List<TranscriptItem>, waiting: Boolean = false): List<ChatBlock> {
        val blocks = mutableListOf<ChatBlock>()
        for (item in items) {
            when (item) {
                is TranscriptItem.User -> blocks += ChatBlock.User(item.id, item.text, item.at, item.attachments)
                is TranscriptItem.Marker -> blocks += ChatBlock.Marker(item.id, item.text, item.at)
                is TranscriptItem.Assistant -> {
                    val reply = item.turn
                    val existing = blocks.lastOrNull() as? ChatBlock.Turn
                    if (existing != null) blocks.removeAt(blocks.lastIndex)
                    val turn = existing?.turn ?: AgentTurn(reply.id, emptyList(), false, reply.at)
                    blocks += ChatBlock.Turn(append(reply, turn).copy(streaming = reply.streaming))
                }
            }
        }
        if (waiting) {
            val last = blocks.lastOrNull()
            if (last is ChatBlock.Turn) {
                blocks[blocks.lastIndex] = ChatBlock.Turn(last.turn.copy(streaming = true))
            } else {
                blocks += ChatBlock.Turn(AgentTurn(PENDING_TURN_ID, emptyList(), true, null))
            }
        }
        return blocks
    }

    private fun append(reply: AssistantTurn, turn: AgentTurn): AgentTurn {
        val segments = turn.segments.toMutableList()
        val steps = mutableListOf<WorkStep>()
        if (reply.thinking.isNotBlank()) steps += WorkStep.Thinking("${reply.id}-thinking", reply.thinking)
        steps += reply.tools.map(WorkStep::Tool)
        val hasText = reply.text.isNotBlank()
        if (steps.isNotEmpty() || hasText) {
            // New output means the attempts that failed before it were retried
            // successfully; like the desktop, those failures are dropped.
            while (segments.lastOrNull() is TurnSegment.Error) segments.removeAt(segments.lastIndex)
        }
        if (steps.isNotEmpty()) {
            val previous = segments.lastOrNull() as? TurnSegment.Work
            if (previous != null) {
                segments[segments.lastIndex] = previous.copy(steps = previous.steps + steps)
            } else {
                segments += TurnSegment.Work("${reply.id}-work", steps)
            }
        }
        if (hasText) segments += TurnSegment.Text("${reply.id}-text", reply.text)
        val error = reply.error
        if (error != null) {
            val previous = segments.lastOrNull() as? TurnSegment.Error
            if (previous != null && previous.message == error) {
                segments[segments.lastIndex] = previous.copy(count = previous.count + 1)
            } else {
                segments += TurnSegment.Error("${reply.id}-error", error, 1)
            }
        }
        return turn.copy(segments = segments)
    }
}
